### ISSUE TITLE
Creating better solvers like LM and Dogleg

### ISSUE THREAD

**Issue by ProfFan** [borglab/SwiftFusion#36]
As illustrated by our attempt in using SwiftFusion for motion planning, I discovered that the NLCG solver cannot deal with a 40-point long dynamic planning problem:
```swift
func planning(_ p_target: Pose2, _ u_all: [[Double]], q0: [Double]) -> [[Double]] {
    var u_final = u_all
    let p0 = fwd_kinematics(q: q0)

    @differentiable
    func e_pose2(_ ŷ: Pose2) -> Double {
        ŷ.t.x * ŷ.t.x + ŷ.t.y * ŷ.t.y
    }

    @differentiable
    func error(_ u: [[Double]]) -> Double {
        var e_total = 0.0
        let e = between(fwd_kinematics(q: u.differentiableReduce(q0,
            { q, u in
                precondition(u.count == 7)
                return (Tensor<Double>(q) + Tensor<Double>(u)).scalars
            }
        )), p_target)
        /// error + control effort
        e_total += e_pose2(e)
        e_total += 400 * u.differentiableReduce(0, { $0+$1.differentiableReduce(0, {e0, v in e0 + v*v })}) // change 400 to anything else to make this boom
        return e_total
    }

    // change 100 to anything else also make this kaboom
    let optimizer = NLCG(for: u_final, max_iteration: 100)

    optimizer.optimize(loss: error, model: &u_final)
    print(error(u_final))
    print(fwd_kinematics(q:(Tensor<Double>(q0) + Tensor<Double>(u_final[0])).scalars))
    // print(proj) // [u0, u1, u2, proj[3], proj[4],proj[5], proj[6]]
    return u_final
}
```

Solving this will probably need LM and dogleg.

**Comment by dellaert** [comment]
Why not? Stuck in local minima?
LM is easy, then CGLS for solve

**Comment by ProfFan** [comment]
@dellaert It diverged...

**Comment by dellaert** [comment]
And, by “it”, you mean?

**Comment by ProfFan** [comment]
I think the result from the manipulation demo and #69 reveals the necessity for a trust region. For example in #69, the CGLS error decreases 10x while the non-linear error explodes.

**Comment by ProfFan** [comment]
Just tried some more datasets. I'll post all experiments to a notebook.

Appears the non-convergence in `sphere_extremenoise` is due to the dataset having too much variation in the noise model that cannot be handled by the current design.

Benchmarks:

- `garage.g2o`: `436365859791.0 ns` (436s)

**Comment by dellaert** [comment]
SE-Sync? What about GTSAM?

**Comment by dellaert** [comment]
PS cool figs in notebook but a bit hard to follow. Consider adding some text cells with explanation.

**Comment by ProfFan** [comment]
> SE-Sync? What about GTSAM?

I found the SE-Sync repo [here](https://github.com/david-m-rosen/SE-Sync) with both dataset and timing, so I just compared against the time on their paper

> PS cool figs in notebook but a bit hard to follow. Consider adding some text cells with explanation.

Thanks for the suggestion! Will do.

**Comment by ProfFan** [comment]
@marcrasi I pushed the very rudimentary WIP to branch [feature/lm](https://github.com/borglab/SwiftFusion/tree/feature/lm), if you have time can you do a profile on `Pose3SLAM`? Profiling with XCode seems to be broken because of https://bugs.swift.org/browse/TF-748

**Comment by ProfFan** [comment]
Here is a timing for the optimization process for the `sphere2500` dataset. It is very clear that we are doing a ton of heap allocation and deallocation.

Brief:
![image](https://user-images.githubusercontent.com/5358374/83317415-72509100-a1fa-11ea-841a-ca5c9457e549.png)

Detail:
![image](https://user-images.githubusercontent.com/5358374/83317401-4cc38780-a1fa-11ea-8cef-507cbeb7ee6e.png)

**Comment by ProfFan** [comment]
@marcrasi When implementing LM there is a damping term implemented previously by adding additional terms in the linearized factor graph:
```
dx = VectorValues()
....

for i in dx.keys {
  var j0 = [lambda * Matrix(eye: dx[i].dimension)]
  var b0 = Vector(zeros: dx[i].dimension)
  damped += JacobianFactor([i], j0, b0)
}
```

However, in the new implementation I found it non-trivial to access the dimensions of the underlying types. Is there a recommended way to implement a similar behavior?

**Comment by marcrasi** [comment]
We'll need some good API for iterating over all the values and doing something with them (such as getting their dimension and constructing a corresponding jacobian). I think this is should be possible, and this is a good use case to help design that, thanks!

But you can also achieve this iteration today, without a nice API. An array of jacobians is quite similar to the array of (`zeroTangentVectors`)[https://github.com/borglab/SwiftFusion/blob/2bad66c21d6f38c4bd98a45b4ce4e38d390e1db3/Sources/SwiftFusion/Inference/VariableAssignments.swift#L113], so imitating that will give you a good way of iterating over all the differentiable values and constructing an array of jacobians.

The problem of constructing the jacobian of the right dimension is a tricky, because [NewJacobianFactor](https://github.com/borglab/SwiftFusion/blob/master/Sources/SwiftFusion/Inference/NewJacobianFactor.swift) needs to know the dimension of its matrix at compiletime, and there's no easy way in Swift to express that the jacobian for a `Vector3` is an `Array3<Vector3>`.

But there's a good way to get around this problem! Define a new type of jacobian factor that instead of storing a matrix just stores one scalar. Then you don't have to compute the dimension of the matrix at compiletime. And this is more efficient too!

I have to go now, but in a few hours I could make a PR implementing this if it would be helpful.

**Comment by ProfFan** [comment]
@marcrasi Thanks for the suggestions! I tried to implement something like this:
```swift
  
  /// Get the tangent vector dimensions
  public var tangentVectorDims: Dictionary<ObjectIdentifier, Int> {
    let r = Dictionary(uniqueKeysWithValues: storage.compactMap {
      (key, value) -> (ObjectIdentifier, Int)? in
      guard let differentiableValue = value.cast(to: AnyDifferentiableStorage.self) else {
        return nil
      }
      return (
        differentiableValue.tangentIdentifier,
        differentiableValue.count
      )
    })
    return r
  }
```
But it seems that `differentiableValue` here is a tree structure?

**Comment by marcrasi** [comment]
Cool, that's pretty close to getting the dimension at runtime. The last step would be to add something like `static var tangentDimension: Int` to [AnyDifferentiableStorage](https://github.com/borglab/SwiftFusion/blob/2bad66c21d6f38c4bd98a45b4ce4e38d390e1db3/Sources/SwiftFusion/Inference/ValuesStorage.swift#L23). There's also a bunch of boilerplate in a few protocols and extensions below that where you have to add `tangentDimension`. (Dave is working on reducing that boilerplate, so hopefully it'll get better soon). You can follow the `var tangentIdentifier: ObjectIdentifier` because the code for dimension should work similarly.

Then `differentiableValue.tangentDimension` will give you the dimension.

However, this won't be enough to construct the jacobians that you need because of the fact that the `NewJacobianFactor` needs to know the type of the fixed-size matrix.

The scalar jacobian idea that I had will address this problem, and it doesn't need to know the dimension at all. I'm back and I can create a quick PR now illustrating this idea.
